core/src/utils/focus-trap.ts

```diff
@@ -0,0 +1,82 @@
+import { focusVisibleElement } from '@utils/helpers';
+
+/**
+ * This query string selects elements that
+ * are eligible to receive focus. We select
+ * interactive elements that meet the following
+ * criteria:
+ * 1. Element does not have a negative tabindex
+ * 2. Element does not have `hidden`
+ * 3. Element does not have `disabled` for non-Ionic components.
+ * 4. Element does not have `disabled` or `disabled="true"` for Ionic components.
+ * Note: We need this distinction because `disabled="false"` is
+ * valid usage for the disabled property on ion-button.
+ */
+export const focusableQueryString =
+  '[tabindex]:not([tabindex^="-"]):not([hidden]):not([disabled]), input:not([type=hidden]):not([tabindex^="-"]):not([hidden]):not([disabled]), textarea:not([tabindex^="-"]):not([hidden]):not([disabled]), button:not([tabindex^="-"]):not([hidden]):not([disabled]), select:not([tabindex^="-"]):not([hidden]):not([disabled]), .ion-focusable:not([tabindex^="-"]):not([hidden]):not([disabled]), .ion-focusable[disabled="false"]:not([tabindex^="-"]):not([hidden])';
+
+/**
+ * Focuses the first descendant in a context
+ * that can receive focus. If none exists,
+ * a fallback element will be focused.
+ * This fallback is typically an ancestor
+ * container such as a menu or overlay so focus does not
+ * leave the container we are trying to trap focus in.
+ */
+export const focusFirstDescendant = <T extends HTMLElement>(ref: Element, fallbackElement: T) => {
```

Comment:
I think we can simplify some of the usage pattern here by using an additional type parameter. 

For example:
```ts
export const focusFirstDescendant = <R extends Element, T extends HTMLElement>(ref: R, fallbackElement: T) => {
  const firstInput = ref.querySelector<HTMLElement>(focusableQueryString);

  focusElementInContext(firstInput, fallbackElement);
};
```

Takes the mark-up that used to be:
```ts
focusFirstDescendant<HTMLIonMenuElement>(el, el);
```

and allows us to do:
```ts
focusFirstDescendant(el, el);
```

The type parameters will be computed based on the type of the arguments passed into the utility function:

![CleanShot 2024-02-27 at 23 45 48@2x](https://github.com/ionic-team/ionic-framework/assets/13732623/1bcb4af5-2006-49ec-beed-995b1afeaf62)

You will also notice that I updated the usage of the `.querySelector` to use a type parameter instead of casting. This is a much safer type inference vs. assigning the type signature manually.

Reply:
I do think it is strange that `ref` and `fallbackElement` are different type signatures, while being a reference to the same element in some cases. Is there a reason why it is `Element` vs. `HTMLElement`?

Reply:
It's possible this was an oversight when originally implemented. For this I tried to minimize refactoring and re-use what we already had as much as possible.

Reply:
Ok I made a couple changes:

1. [Converged on `HTMLElement` instead of `Element` and `HTMLElement`](https://github.com/ionic-team/ionic-framework/pull/29091/commits/3f9f94c2416e6d6bbf9e65573687938208e06f4f).
2. [Added additional type parameter as you suggested.](https://github.com/ionic-team/ionic-framework/pull/29091/commits/37ad77a304fdbceb47fd6c5f2888bc7897f48d1c)